Tests/NIOCoreTests/ByteBufferTest.swift

```diff
@@ -1891,7 +1891,24 @@ class ByteBufferTest: XCTestCase {
         XCTAssertTrue(buffer.shrinkBufferCapacity(to: desiredCapacity))
         XCTAssertEqual(buffer.capacity, 1024)
     }
-    
+
+    func testShrinkBufferCapacityWithLeadingUnwrittenBytes() {
+        var buffer = self.allocator.buffer(capacity: 16384)
+        buffer.moveWriterIndex(to: 16000)
+        buffer.moveReaderIndex(to: 16000)
+        buffer.writeString("WW")
+        buffer.shrinkBufferCapacity(to: 4)
+        XCTAssertEqual("WW", String(buffer: buffer))
+
+        // If readable bytes is exactly the same as buffer capacity shrunken to
+        buffer = self.allocator.buffer(capacity: 16)
+        buffer.moveWriterIndex(to: 8)
+        buffer.moveReaderIndex(to: 8)
+        buffer.writeString("WWWWWWWW")  // 8 bytes written
+        buffer.shrinkBufferCapacity(to: 4)  // Invisible padding makes this 8 bytes
+        XCTAssertEqual("WWWWWWWW", String(buffer: buffer))  // All 8 bytes are returned!
```

Comment:
Can we very specifically look at the 2nd test in `testShrinkBufferCapacityWithLeadingUnwrittenBytes` and the comments and make sure it is what is expected. Or if I am testing implementation detail that should not fail if the function changes?